File: algorithm/python/BOJ15654.py
# from https://www.acmicpc.net/problem/15654

import sys
N, M = map(int, sys.stdin.readline().split())
input_list = list(map(int, sys.stdin.readline().split()))
input_list.sort()
output_list = [0] * N
is_selected = [False] * N


def permutation(cnt):
    if cnt == M:
        for i in range(0, M):
            print(output_list[i], end=" ")
        print()
        return

    for i in range(0, N):
        if is_selected[i]:
            continue
        is_selected[i] = True
        output_list[cnt] = input_list[i]
        permutation(cnt + 1)
        is_selected[i] = False


permutation(0)

File: algorithm/typescript/BOJ15654.ts
// from https://www.acmicpc.net/problem/15654

import fs = require("fs")

const str: string[] = fs.readFileSync("/dev/stdin").toString().split("\n");
const NM = str[0].split(" ");
const N: number = parseInt(NM[0]);
const M: number = parseInt(NM[1]);

const input = str[1].split(" ").map(Number);
input.sort(function(a, b) { return a - b });
const output = [];
const result = [];
const isSelected = Array(N).fill(false);

function permutation(cnt: number): void {
  if (cnt == M) {
    for (let i = 0; i < M; i++) {
      result.push(`${output[i]} `);
    }
    result.push(`\n`);
    return;
  }

  for (let i = 0; i < N; i++) {
    if (isSelected[i]) continue;
    isSelected[i] = true;
    output[cnt] = input[i];
    permutation(cnt + 1);
    isSelected[i] = false;
  }
}

permutation(0);
console.log(result.join(''));
